eni: fix new node not immediately triggers the creation of Eni before instances api  sync

wu0407 opened the pull request:
Update the instance network interface information when a new node is created.
This is fix errors return by `ResyncInterfacesAndIPs` "Instance not found! Please delete
corresponding ciliumnode if instance has already been deleted" when new node
created in cloud provider then add to cluster and before resync triggered.

Fixes: #20678

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

```release-note
Eliminate a delay between new node creation and creation of ENIs
```

gandro commented on pkg/ipam/types/types.go:
```diff
@@ -482,3 +482,20 @@ func (m *InstanceMap) NumInstances() (size int) {
 }
+
+// IsExist returns whether the instance ID is in the instanceMap
+func (m *InstanceMap) IsExist(instanceID string) (exist bool) {
```
nit:

```suggestion
// Exists returns whether the instance ID is in the instanceMap
func (m *InstanceMap) Exists(instanceID string) (exists bool) {
```

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -115,6 +115,12 @@ type AllocationImplementation interface {
+	// UpdateInstance is sync interfaces info for new instance in instances if new instance not exist
+	UpdateInstance(ctx context.Context, obj *v2.CiliumNode) time.Time
+
+	// DeleteInstance is delete instance form instances
```
```suggestion
	// DeleteInstance deletes the instance from instances
```

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -279,7 +282,13 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
+		node.ops, instanceExist = n.instancesAPI.CreateNode(resource, node)
+		if !instanceExist {
+			if _, ok := n.instancesAPIResync(context.TODO()); !ok {
+				node.logger().Warning("Failed to resync instance API after found new node")
```
```suggestion
				node.logger().Warning("Failed to resync instance API after new node was found")
```

gandro commented on pkg/ipam/node_manager.go:
```diff
@@ -347,7 +356,13 @@ func (n *NodeManager) Delete(nodeName string) {
 
-	delete(n.nodes, nodeName)
+	// Delete instance from instanceManager in case Node.recalculate() use stale data
+	// when nodes from being deleted and then rejoined into the cluster
```
Can you elaborate on that comment? I don't see how removing the node from the instances list affects `n.recalcuate()`. Can you point me to the code which will behave differently? Thanks!

wu0407 replied:
Such as I want downgrade EC2 instance configuration，so remove node from cluster and change instance type, then rejoined in the cluster. The number of instance bound ENIs may exceed the limit, it need unbound ENI. The instance of enis is changed. If the node rejoins the cluster before next `instanceAPIResync`,  `ResyncInterfacesAndIPs` return incorrect result.

gandro replied:
Ah, so this is relying on the code added by this PR (which forces an immediate resync if the instance is not in the list). Makes sense. 

Let's word it a little bit differently though, because this wasn't clear to me just from the comment:

```suggestion
	// Delete the instance from instanceManager. This will cause Update() to
	// invoke instancesAPIResync if this instance rejoins the cluster.
	// This ensures that Node.recalculate() does not use stale data for
	// instances which rejoin the cluster after their EC2 configuration has changed.
```

gandro reviewed (changes requested):
Thanks for the updated version. I didn't realize that we can't call `instancesAPIResync` from the `InstancesManager`. So maybe keeping `CreateNode` is it was previously might be cleaner.

I think we're getting close though. Looking at the code again, I feel the the following might not be terrible:

First, let's undo the change to `CreateNode` again (sorry for that). Instead, let's change the `AllocationImplementation` to have a `HasInstance` and `DeleteInstance` method. The `DeleteInstance` is already there, the `HasInstance` checks if the instance exists and returns a boolean.

In `NodeManager.Update`, let's then trigger the resync if the detect that an instance is missing, for example:

```golang
if !n.instancesAPI.HasInstance(resource.InstanceID()) {
	if _, ok := n.instancesAPIResync(context.TODO()); !ok {
		node.logger().Warning("Failed to resync instance API after new node was found")
	}
}

node.ops = n.instancesAPI.CreateNode(resource, node)
```

I also added some inline comments regarding language.

gandro reviewed (commented):
Looking good! One last suggestion here: https://github.com/cilium/cilium/pull/21027#discussion_r957231113

gandro reviewed (approved):
Thank you!

joestringer reviewed (approved):
Approving on behalf of @cilium/alibabacloud

christarazi commented on pkg/ipam/node_manager.go:
```diff
@@ -279,6 +285,12 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
 
+		if !n.instancesAPI.HasInstance(resource.InstanceID()) {
+			if _, ok := n.instancesAPIResync(context.TODO()); !ok {
+				node.logger().Warning("Failed to resync instance API after new node was found")
```
```suggestion
				node.logger().Warning("Failed to resync the instances from the API after new node was found")
```

wu0407 replied:
> Thanks for the fix! Could you update the commit msg to be more descriptive on what the fix is doing and why?

The updated commit message
```
eni: fix new node not immediately triggers the creation of ENI

1. Sync the instances network interface information from instances api when a new node is created
but not in `instanceManager.instances`. This is fix errors in `ResyncInterfacesAndIPs` "Instance not found! Please delete
corresponding ciliumnode if instance has already been deleted" when new node
created in cloud provider then add to cluster and before resync triggered.

2. Delete instance from `instanceManager.instances` on node deleted. This will cause `NodeManager.Update()` to
invoke `instancesAPIResync` if this instance rejoins the cluster.
This ensures that `Node.recalculate()` does not use stale data for
instances which rejoin the cluster after their EC2 configuration has changed.

Fixes: #20678
```

christarazi reviewed (commented):
Thanks for the fix! Could you update the commit msg to be more descriptive on what the fix is doing and why?

wu0407 committed:
eni: fix new node not triggering creation of ENI

1. Sync the instances network interface information from instances api when a
new node is created but not in `instanceManager.instances`. This is fix errors
 in `ResyncInterfacesAndIPs` "Instance not found! Please delete corresponding
ciliumnode if instance has already been deleted" when new node created in
cloud provider then add to cluster and before resync triggered.

2. Delete instance from `instanceManager.instances` on node deleted. This will
cause `NodeManager.Update()` to invoke `instancesAPIResync` if this instance
rejoins the cluster. This ensures that `Node.recalculate()` does not use stale
data for instances which rejoin the cluster after their EC2 configuration has changed.

Fixes: #20678

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>
